### ISSUE TITLE
TS7023 false positive: contextual return type ignored when a recursive call is assigned to a union-annotated local

### ISSUE THREAD

**Issue by ljharb** [microsoft/typescript-go#4768]
### 🔎 Search Terms

TS7023, recursive function expression, contextual type, contextual signature, implicitly has return type any, union type annotation

### 🕗 Version & Regression Information

- **Fails**: `typescript@7.0.2` (stable) and `typescript@7.1.0-dev.20260724.1`
- **Works**: `typescript@6.0.3` and earlier (strada)

### 💻 Code

```ts
const h: (value?: unknown, other?: unknown) => string | false = function whyH(value, other) {
	if (value === other) { return ''; }
	let equal: string | false = '';
	equal = whyH(value, other);
	return equal;
};
```

Compiled with `--strict --noEmit`.

The same failure occurs in a `checkJs` JSDoc equivalent (`/** @type {(value?: unknown, other?: unknown) => string | false} */` on a `module.exports` assignment), which is where we hit it in the wild: [`inspect-js/is-equal`'s `why.js`](https://github.com/inspect-js/is-equal/blob/main/why.js), whose `whyNotEqual` is contextually typed by its own `.d.ts` via `@type {typeof import('./why')}`. Generic contextual signatures reproduce it identically.

### 🙁 Actual behavior

```
error TS7023: 'whyH' implicitly has return type 'any' because it does not have a return type annotation and is referenced directly or indirectly in one of its return expressions.
```

### 🙂 Expected behavior

No error. The function expression is contextually typed by the variable's type annotation, so its return type is known to be `string | false` — there is no inference cycle to report. TS ≤ 6.0.3 accepts this code.

### Additional notes

- Removing the (redundant but correct) `string | false` annotation on `equal` — i.e. `let equal = '';` — makes the error disappear, suggesting the assignment of the recursive call to a union-annotated local is what spuriously trips the circularity detection.
- Returning the recursive call directly (`return whyH(value, other);`) does not trigger the error; only the assign-to-annotated-local shape does.

**Comment by ahejlsberg** [comment]
I can't reproduce the stated behavior. The repro consistently errors with TS 7.0.2, TS 6.0.3, and TS 5.9.3.

The existence of a contextual type doesn't prevent the compiler from resolving actual expression types--it merely provides context for things like literal types and inferred parameter types for function expressions. So, a circularity in return type inference will still occur in the presence of a contextual type.

**Comment by ahejlsberg** [comment]
BTW, the reason the type annotation on `equal` causes the circularity is that control flow analysis tries to narrow its union type based on the type returned by `whyH`. Without a type annotation this doesn't happen.

**Comment by ljharb** [comment]
This issue was filed by Claude (unbeknownst to me initially) - apologies for that. Here's the session's response, fwiw:

-----
Apologies — the TS-syntax snippet was over-reduced on my end, and you're right that it errors on 5.9/6.0 as well; the "works in 6.0.3" claim doesn't hold for it.

The delta I'm actually seeing is confined to checkJs:

```js
/** @type {(value?: unknown, other?: unknown) => string | false} */
var h = function whyH(value, other) {
	if (value === other) { return ''; }
	/** @type {string | false} */
	var equal = '';
	equal = whyH(value, other);
	return equal;
};
```

With `--allowJs --checkJs --noEmit --strict`:
- 5.9.3: no error
- 6.0.3: no error
- 7.0.2 / 7.1.0-dev.20260724.1: TS7023

The CFA-narrowing explanation makes sense for the TS-syntax version. The question is whether the strada behavior in `.js` files — where the JSDoc `@type` on the declaration evidently supplied the function's type outright, so no return-type inference cycle arose — was dropped intentionally. If this is deliberate alignment of JS checking with TS semantics, that's reasonable, but it may deserve a note in the migration docs: `/** @type {typeof import('./x')} */ module.exports = function ...` self-typed CJS modules are a common JSDoc pattern, and that's where this surfaced.

**Comment by RyanCavanaugh** [comment]
CHANGES.md already describes JSDoc support as now being much more based on TS. Changes that align to TS's behavior basically do not require additional justification -- the goal is that they behave the same unless there's a strong positive case to be made otherwise.

**Comment by ljharb** [comment]
I'm confused - you're saying that my example failed in TS in TS <= 6 but worked in JS, and in TS7, it now fails in both?

That seems like a regression in JS, not an improvement.

**Comment by RyanCavanaugh** [comment]
> That seems like a regression in JS, not an improvement.

Correct. Consistency with TS is the primary goal in JS support going forward - we don't want to maintain two different sets of semantics.
